packages/algob/src/builtin-tasks/compile.ts

```diff
@@ -1,13 +1,112 @@
+import type { Algodv2, CompileOut } from "algosdk";
+import { readdirSync, readFileSync, writeFileSync } from 'fs';
+import * as path from 'path';
+import YAML from "yaml";
+
 import { task } from "../internal/core/config/config-env";
-// import { createClient } from "../lib/driver";
-// import { AlgobRuntimeEnv, TaskArguments } from "../types";
+import { parseAlgorandError } from "../internal/core/errors";
+import { assertDir, ASSETS_DIR, CACHE_DIR } from "../internal/core/project-structure";
+import { createClient } from "../lib/driver";
+import type { AlgobRuntimeEnv, ASCCache, Network } from "../types";
 import { TASK_COMPILE } from "./task-names";
+const murmurhash = require('murmurhash'); // eslint-disable-line @typescript-eslint/no-var-requires
+
+const tealExt = ".teal";
 
 export default function (): void {
-  task(TASK_COMPILE, "Compilation task")
-    .setAction(
-      async () => {
-        console.log("compiling......");
-      }
-    );
+  task(TASK_COMPILE, "Compile all TEAL smart contracts")
+    .addFlag("force", "recompile even if the source file didn't change")
+    .setAction(_compile);
+}
+
+export interface TaskArgs {
+  force: boolean
+}
+
+async function _compile ({ force }: TaskArgs, env: AlgobRuntimeEnv): Promise<void> {
+  const op = new CompileOp(env.network);
+  return await compile(force, op);
+}
+
+export async function compile (force: boolean, op: CompileOp): Promise<void> {
+  await assertDir(CACHE_DIR);
+  const cache = readArtifacts(CACHE_DIR);
+
+  for (const f of readdirSync(ASSETS_DIR)) {
+    if (!f.endsWith(tealExt)) { continue; }
+
+    let c = cache.get(f);
+    const [teal, thash] = readTealAndHash(path.join(ASSETS_DIR, f));
+    if (!force && c !== undefined && c.srcHash === thash) {
+      console.log(`smart-contract source "${f}" didn't change, skipping.`);
+      continue;
+    }
+    console.log("compiling", f);
+    c = await compileAndSave(op, new ASCCachePartial(f, teal, thash));
+    cache.set(f, c);
```

Comment:
You only use `c.srcHash` from this type in `c.srcHash === thash`. So all of other data is not needed in the type.
Timestamp is not checked as well so either it should be checked or removed.

I also think that `const [teal, thash] = readTealAndHash(...)` already directly returns all that you care about: the hash.
So you should only index by `filename -> hash` or `filename -> {hash, timestamp}` and then remove this type and its interface.

Reply:
It is used in the next line - we serialize the whole `c: ASCCache` structure.

Reply:
I know but you only used the hash and you didn't use any of the other variables. So they are not needed and can be added when we'll decide to use them.
Also it's strange that you create `ASCCachePartial` when instead you could return it from `readTealAndHash` and not create a tuple there. So either the tuple `[teal, thash]` or this type should be changed.

Reply:
You replaced the type. So it doesn't matter.